### DISCUSSION TITLE
create-next-app with ESM modules

### DISCUSSION THREAD

**Discussion by sigma54-ca** [vercel/next.js#32901] [Help] [answered]
As of v12, nextJS is supposed to support ESM modules. However, I'm having trouble getting it to work with a basic application created via create-next-app (Node v14.18.1) 

Here are the steps I'm taking:
1) Run "npx create-next-app" with no additional options. nextJS v12.0.7 is installed with default template.
2) Add "type": "module" to package.json
3) Update next.config.js to:
```
 /**
 * @type {import('next').NextConfig}
 */
const nextConfig = {
  /* config options here */
}

export default nextConfig
```
4) Run "npm run-script dev"
5) In the console:
 
> ready - started server on 0.0.0.0:3000, url: http://localhost:3000
> warn  - Detected next.config.js, no exported configuration found. https://nextjs.org/docs/messages/empty-configuration
> event - compiled client and server successfully in 1740 ms (169 modules)

6) As soon as I go to localhost:3000, the following appears in the console:
> wait  - compiling / (client and server)...
> event - compiled client and server successfully in 437 ms (187 modules)
> wait  - compiling /_error (client and server)...
> event - compiled client and server successfully in 68 ms (188 modules)
> Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: C:\Data\Programming\myapp\.next\server\pages\_document.js
> require() of ES modules is not supported.
> require() of C:\Data\Programming\myapp\.next\server\pages\_document.js from C:\Data\Programming\myapp\node_modules\next\dist\server\require.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
> Instead rename _document.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from C:\Data\Programming\myapp\package.json.
> 
>     at new NodeError (internal/errors.js:322:7)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1102:13)
>     at Module.load (internal/modules/cjs/loader.js:950:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:790:12)
>     at Module.require (internal/modules/cjs/loader.js:974:19)
>     at require (internal/modules/cjs/helpers.js:93:18)
>     at Object.requirePage (C:\Data\Programming\myapp\node_modules\next\dist\server\require.js:48:12)
>     at Object.loadComponents (C:\Data\Programming\myapp\node_modules\next\dist\server\load-components.js:56:23)
>     at DevServer.findPageComponents (C:\Data\Programming\myapp\node_modules\next\dist\server\next-server.js:1243:63)
>     at DevServer.findPageComponents (C:\Data\Programming\myapp\node_modules\next\dist\server\dev\next-dev-server.js:666:26) {
>   code: 'ERR_REQUIRE_ESM'
> }
> error - Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: C:\Data\Programming\myapp\.next\server\pages\_document.js
> require() of ES modules is not supported.
> require() of C:\Data\Programming\myapp\.next\server\pages\_document.js from C:\Data\Programming\myapp\node_modules\next\dist\server\require.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
> Instead rename _document.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from C:\Data\Programming\myapp\package.json.
> 
> Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: C:\Data\Programming\myapp\.next\server\pages\_document.js
> require() of ES modules is not supported.
> require() of C:\Data\Programming\myapp\.next\server\pages\_document.js from C:\Data\Programming\myapp\node_modules\next\dist\server\require.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
> Instead rename _document.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from C:\Data\Programming\myapp\package.json.
> 
>     at new NodeError (internal/errors.js:322:7)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1102:13)
>     at Module.load (internal/modules/cjs/loader.js:950:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:790:12)
>     at Module.require (internal/modules/cjs/loader.js:974:19)
>     at require (internal/modules/cjs/helpers.js:93:18)
>     at Object.requirePage (C:\Data\Programming\myapp\node_modules\next\dist\server\require.js:48:12)
>     at Object.loadComponents (C:\Data\Programming\myapp\node_modules\next\dist\server\load-components.js:56:23)
>     at DevServer.findPageComponents (C:\Data\Programming\myapp\node_modules\next\dist\server\next-server.js:1243:63)
>     at DevServer.findPageComponents (C:\Data\Programming\myapp\node_modules\next\dist\server\dev\next-dev-server.js:666:26) {
>   code: 'ERR_REQUIRE_ESM'
> }

Is there a way to get nextJS v12 to work with type "module" declared in the package JSON file? Thanks!

**Comment by balazsorban44** [comment, marked as answer]
To clarify, ESM support refers to the input and output of Next.js, so modules/packages you would use _within_ the app.

https://nextjs.org/blog/next-11-1#es-modules-support

https://nextjs.org/blog/next-12#es-modules-support-and-url-imports

What would be the reason for you needing app-level ESM support?

**Reply by sigma54-ca** [reply]
I was able to get the application to compile and run by removing the type: module from my package.json file.

However, when trying to run tests with ts-mocha or when trying to run a debug script, I keep getting: "SyntaxError: Cannot use import statement outside a module". It suggests modifying the package.json or changing the extension to mjs.  I can't change the extension to mjs since I'm using Typescript. I tried using a debug tsconfig (as shown below) that sets the module type to CommonJs but that doesn't seem to work either. 

`{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "module": "CommonJS"
  }
}
`

I think there is some way to override the default node loader to use esm to be able to run those but I haven't figured out a way to get it to work yet.

**Reply by inwardmovement** [reply]
Is it still not possible to use app-level ESM?
